### PR TITLE
Multiarch image adjustments

### PR DESCRIPTION
**Pull request by frozenprocess**
Fix: This PR updates all non `amd64` dockerfiles for a multiarch release.

```release-note
None required
```

### CODE REVIEW

File: Dockerfile.arm64 [hunk 1]
```diff
@@ -35,6 +35,8 @@ LABEL name="Calico CLI tool" \
 
 COPY --from=ubi /licenses /licenses
 ADD bin/calicoctl-linux-arm64 /calicoctl
+# Overwrite qemu binary
+ADD LICENSE /usr/bin/qemu-aarch64-static
```
Comment (frozenprocess) [inline Dockerfile.arm64:39]: At this point our `arm64` image doesn't have executables to invoke normal commands for removing `qemu-aarch64-static` binary, so I just overwrite the file with a text file.

File: Makefile [hunk 2]
```diff
@@ -138,7 +140,7 @@ remote-deps: mod-download
 image: $(CALICOCTL_IMAGE)
 $(CALICOCTL_IMAGE): $(CTL_CONTAINER_CREATED)
 $(CTL_CONTAINER_CREATED): register Dockerfile.$(ARCH) bin/calicoctl-linux-$(ARCH)
-	docker build -t $(CALICOCTL_IMAGE):latest-$(ARCH) $(TARGET_PLATFORM) --build-arg QEMU_IMAGE=$(CALICO_BUILD) --build-arg GIT_VERSION=$(GIT_VERSION) -f Dockerfile.$(ARCH) .
+	docker build --pull -t $(CALICOCTL_IMAGE):latest-$(ARCH) $(TARGET_PLATFORM) --build-arg QEMU_IMAGE=$(CALICO_BUILD) --build-arg GIT_VERSION=$(GIT_VERSION) -f Dockerfile.$(ARCH) .
```
Comment (frozenprocess) [inline Makefile:143]: `--pull`  = Always attempt to pull a newer version of the image

File: Makefile [hunk 3]
```diff
@@ -49,6 +49,8 @@ Makefile.common.$(MAKE_BRANCH):
 include Makefile.common
 
 ifeq ($(ARCH),arm64)
+# Forces ARM64 build image to be used in a crosscompilation run.
+CALICO_BUILD:=$(CALICO_BUILD)-$(ARCH)
```
Comment (frozenprocess) [inline Makefile:53]: When docker pulls `go-build:v0.55` for `arm64` it tags it incorrectly and use it for all other architectures which causes problems.
In this line we make sure that image pulled by docker for `arm64` process is tagged correctly in local repo. `go-build:v0.55-arm64`

File: Dockerfile.arm64 [hunk 4]
```diff
@@ -11,8 +11,8 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-ARG QEMU_IMAGE=calico/go-build:v0.55
-FROM ${QEMU_IMAGE} as qemu
+ARG QEMU_IMAGE
+FROM ${QEMU_IMAGE}-arm64 as qemu
```
Comment (frozenprocess) [inline Dockerfile.arm64:15]: In order to tag the image with correct achitecture we are using `--platform` with the `docker build` command causing docker to pull the `arm64` image to the local repository and tag it as `v0.55` then docker uses this localy stored image to build all other architectures causing an issue in binary build step.

File: Dockerfile.s390x [hunk 5]
```diff
@@ -1,4 +1,9 @@
-FROM s390x/alpine:3.10
+ARG QEMU_IMAGE
+FROM ${QEMU_IMAGE} as qemu
+
+FROM s390x/alpine:3.12 as base
```
Comment (neiljerram) [inline Dockerfile.s390x:4]: Do we need "as base" here?  Please remove if not.

File: Dockerfile.ppc64le [hunk 6]
```diff
@@ -1,4 +1,9 @@
-FROM ppc64le/alpine:3.10
+ARG QEMU_IMAGE
+FROM ${QEMU_IMAGE} as qemu
+
+FROM ppc64le/alpine:3.12 as base
```
Comment (neiljerram) [inline Dockerfile.ppc64le:4]: Do we need "as base" here? Please remove if not.

File: Dockerfile.armv7 [hunk 7]
```diff
@@ -1,5 +1,10 @@
-FROM arm32v7/alpine:3.12
-MAINTAINER Marc Crébassa <aalaesar@gmail.com>
+ARG QEMU_IMAGE
+FROM ${QEMU_IMAGE} as qemu
+
+FROM arm32v7/alpine:3.12 as base
```
Comment (neiljerram) [inline Dockerfile.armv7:4]: Do we need "as base" here? Please remove if not.

File: Dockerfile.armv7 [hunk 8]
```diff
@@ -1,5 +1,10 @@
...
+COPY --from=qemu /usr/bin/qemu-arm-static /usr/bin/
+
+MAINTAINER LoZ Open Source Ecosystem (https://www.ibm.com/developerworks/community/groups/community/lozopensource)
```
Comment (neiljerram) [inline Dockerfile.armv7:7]: Why this change?  Probably better to have a separate PR for this, so that the reason for the change is clearly described.

File: Dockerfile.arm64 (hunk 1 above)
Comment (neiljerram) [reply Dockerfile.arm64:39]: How exactly does that help?  Who is it that checks for the existence of /usr/bin/qemu-aarch64-static and then tries to run it?  With this solution, will they still try to run it, and then hit an error because it's just plain text?

File: Dockerfile.arm64 (hunk 4 above)
Comment (neiljerram) [reply Dockerfile.arm64:15]: I'm sorry, I don't understand.  Why do we need to add "-arm64" here, but we don't do that for the other architectures?

> In order to tag the image

Which image do you mean?

> with correct architecture we are using --platform with the docker build command causing docker to pull the arm64 image

Which image do you mean here?

> to the local repository and tag it as v0.55

Huh?  It should only be calico-build that has the v0.55 tag, but I don't think you mean the calico-build image here, do you?

> then docker uses this locally stored image to build all other architectures causing an issue in binary build step.

**Review by neiljerram — CHANGES_REQUESTED** [review]

File: Dockerfile.arm64 (hunk 4 above)
Comment (frozenprocess) [reply Dockerfile.arm64:15]: > Why do we need to add "-arm64" here, but we don't do that for the other architectures?

It is because of the `--platform=linux/arm64/v8` pulls an `arm64` image and tags it as both `calico/go-build:v0.55` and `calico/go-build:v0.55-arm64`.


In `arm64` dockerfile `FROM SCRATCH` pulls an empty `amd64` layer in the build process which results in an incorrect architecture tag for the output image (aka `calico/ctl`).`--platform` prevents docker from tagging image (`calico/ctl`) architecture as `amd64`.

Here is an example from cni repo:
https://github.com/projectcalico/cni-plugin/blob/f6cd1b86daec16b04e095b04abf0af3bffb73a89/Dockerfile.arm64#L1
```
docker inspect calico/cni:v3.20.1-arm64 | egrep -i arch
        "Architecture": "amd64",
```

> Which image do you mean?

Output image aka `calico/ctl:vxxxxx-arm64`


> Which image do you mean here?

`calico/go-build:v0.55-arm64` will be pulled and stored as both `calico/go-build:v0.55` and `calico/go-build:v0.55-arm64`

> Huh? It should only be calico-build that has the v0.55 tag, but I don't think you mean the calico-build image here, do you?

No, I meant `calico/go-build:v0.55-arm64` will be saved as `calico/go-build:v0.55`
```
root@build-dummy:~/calicoctl# docker image ls
REPOSITORY   TAG       IMAGE ID   CREATED   SIZE
root@build-dummy:~/calicoctl# docker pull --platform=linux/arm64/v8 calico/go-build:v0.55 &> /dev/null
root@build-dummy:~/calicoctl# docker image ls
REPOSITORY        TAG       IMAGE ID       CREATED      SIZE
calico/go-build   v0.55     633df77ba924   5 days ago   1.74GB
root@build-dummy:~/calicoctl# docker inspect 633df77ba924 | egrep -i arch
        "Architecture": "arm64",
```

File: Dockerfile.s390x (hunk 5 above)
Comment (frozenprocess) [reply Dockerfile.s390x:4]: We don't need it and It can be removed.

File: Dockerfile.armv7 (hunk 8 above)
Comment (frozenprocess) [reply Dockerfile.armv7:7]: Since `arm32v7/alpine` is an `arm` based image and in it executable files are `arm` only we neeed the qemu binary to run the commands like the following one
https://github.com/projectcalico/calicoctl/blob/136af0e96783676c7faba39c71b1ebf390d04dd6/Dockerfile.armv7#L9

**Review by caseydavenport — APPROVED** [review]
Alright this has been broken long enough. Up up and away!
